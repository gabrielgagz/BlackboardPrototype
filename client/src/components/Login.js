import { useState, useEffect, useRef } from 'react';
import  { useHistory, useLocation } from 'react-router-dom'
import { useSocket } from '../hooks/useSocket';
import QRCode from 'qrcode';
import { toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import '../css/login.css';

export const Login = () => {

    // Start with an empty socket
    const [token, setToken] = useState();

    // Generator button state
    const [state, setState] = useState( false );

    // Socket connection
    const { events } = useSocket( token );

    // Save generator button ref
    const button = useRef();

    // Use Router history
    const history = useHistory();

    const location = useLocation();

    useEffect( () => {

        if (location.state) {
            
        // Show toast error when you open a third session in the same room 
            if ( Object.keys(location.state).length > 0 ) {
                
                toast.error('User limit was exceeded!', {
                    position: "bottom-center",
                    autoClose: 2000,
                    hideProgressBar: true,
                    closeOnClick: false,
                    pauseOnHover: true,
                    draggable: true,
                    progress: undefined,
                });
            
                history.replace( {
                    pathname: location.pathname,
                    state: {}
                } );
            
            }
        }

        // Check if we have at least one event
        if ( events.length > 0 ) {

            if ( events[0][0].token === token ) {
                
                // Disable button and redirect to blackboard component
                button.current.disabled = true;
                history.push(`/login/${ token }/web`);
            }
        }

    },[ events ]);

    const generateToken = () => {

        // Token length
        const length = 16;

        const a = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890".split("");
        let b = [];  

        for (let i=0; i<length; i++) {

        const j = (Math.random() * (a.length-1)).toFixed(0);
        b[i] = a[j];

        }

        // Qr call
        generateQR(b.join("")  )

        // Return new token
        return b.join("");
    }

    const generateQR = ( token ) => {

        button.current = document.querySelector('.btn');
        const loginCanvas = document.querySelector('#loginCanvas');

        const url = window.location.href + `login/${ token }/qr`;

        // Generate Qr
        const setQR = async () => {

        try {
            await QRCode.toCanvas( loginCanvas, url, { scale: 7     } );
        } catch (err) {
            console.error(err);
        }

        }

        // Change text if button is clicked once
        if ( !state ) {
            button.current.innerHTML = "RELOAD QR";
            setState( true );
        }

        // Generate Qr
        setQR();
    }

    return (
        <>
            <button 
            className='btn btn-login btn-lg btn-success mt-3 py-2 px-3'
            onClick={ () => setToken( generateToken() ) }
            >
                GENERATE QR
            </button>
            <canvas id='loginCanvas'></canvas>
        </>
    )

}



