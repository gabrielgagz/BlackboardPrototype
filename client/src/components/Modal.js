import React from "react";
import { useHistory } from "react-router-dom";

export const Modal = ( modalAction ) => {

    // Use Router history
    const history = useHistory();

    const executeAction = () => {
        history.push("/");
    };

    return (
        <div
            className="modal fade"
            id="bbModal"
            data-bs-backdrop="static"
            data-bs-keyboard="false"
            tabIndex="-1"
            aria-labelledby="bbModalLabel"
            aria-hidden="true"
        >
            <div className="modal-dialog">
                <div className="modal-content">
                    <div className="modal-body">
                    {
                    ( modalAction ) ?
                    'Are you sure you want to quit?'
                    : 'User limit (2) was exceeded for this session.'
                    }
                    </div>
                    <div className="modal-footer">
                        {
                        ( modalAction ) &&
                        <button
                            type="button"
                            className="btn btn-secondary"
                            data-bs-dismiss="modal"
                        >
                            Close
                        </button>
                        }
                        <button type="button" className="btn btn-danger"
                        data-bs-dismiss="modal"
                        onClick={ executeAction }
                        >
                            Quit
                        </button>
                    </div>
                </div>
            </div>
        </div>
    );
};
