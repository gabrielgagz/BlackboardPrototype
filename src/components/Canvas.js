import React, { useEffect } from "react";
import { useSocket } from "../hooks/useSocket";
import PropTypes from 'prop-types';
import "../css/canvas.css";


export const Canvas = ( uuid ) => {

    const urlToken = window.location.pathname.replace("/blackboard/", "");

    // Importamos el Hook para el uso de WebSocket
    const { draws, sendDraw } = useSocket( urlToken );

    useEffect(() => {

        // Definimos el entorno
        const canvas = document.querySelector("#bCanvas");
        const context = canvas.getContext("2d");
        let lastPoint;

        // Tamaño de la ventana
        canvas.width = window.innerWidth;
        canvas.height = window.innerHeight;

        // Definimos el trazado
        context.strokeStyle = 'white';
        context.lineWidth = 4;
        context.lineCap = 'round';

        // Escribimos en la pizarra
        const draw = ( data ) => {

            // Obtenemos el trazado del servidor
            const { lastPoint, x, y } = data[0];

            context.beginPath();
            context.moveTo(lastPoint.x, lastPoint.y);
            context.lineTo(x, y);
            context.stroke();
            context.closePath();

        }
        

        // Acctiones al mover el mouse
        // TODO: Implementar touch input
        const drawMove = (e) => {

            if ( e.buttons ) {

                if (!lastPoint) {
                    lastPoint = { x: e.offsetX, y: e.offsetY };
                    return;
                }

                sendDraw([{
                    lastPoint, 
                    x: e.offsetX,
                    y: e.offsetY
                }]);
        
                lastPoint = { x: e.offsetX, y: e.offsetY };
                
            }

        }

        const drawUp = () => {

            lastPoint = undefined;
            
        }

        // Recibimos los datos del servidor 
        // los y dibujamos en pantalla
        draws.forEach( draw );

        // Implemetación de touch screen
        const touchmove = ( e ) => {

            drawMove(e.changedTouches[0]);
            e.preventDefault(); 

        } 

        canvas.onmousemove = drawMove;
        canvas.onmouseup = drawUp;
        
    },);

    return <canvas id="bCanvas"></canvas>;
};

Canvas.propTypes = {
    uuid: PropTypes.string
};