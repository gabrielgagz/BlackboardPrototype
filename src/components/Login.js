import { useState, useEffect, useRef } from 'react';
import  { useHistory } from 'react-router-dom'
import { useSocket } from '../hooks/useSocket';
import QRCode from 'qrcode';
import '../css/login.css';


export const Login = () => {

    // Iniciamos el token vacio
    const [token, setToken] = useState();

    // State del boton generador
    const [state, setState] = useState( false );

    // Conectamos al Socket
    const { events } = useSocket( token );

    // Guardamos la referencia al boton generador
    const button = useRef();

    // Usamos history del Router
    const history = useHistory();

    // Chequeamos que events cambie. El primer cambio corresponderá a la carga a través del QR nos redirigimos a la pizarra con el token correspondiente
    useEffect( () => {

        // Si hay evento, ejecutamos las acciones para la redirección
        if ( events.length > 0 ) {

            if ( events[0][0].token === token ) {
                
                // Deshabilitamos el boton y redireccionamos a la pizarra

                button.current.disabled = true;
                
                history.push(`/login/${ token }`)
            }
        }

    },[ events ]);

    const generateToken = () => {

        // Longitud del token
        const length = 16;

        const a = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890".split("");
        let b = [];  

        for (let i=0; i<length; i++) {

        const j = (Math.random() * (a.length-1)).toFixed(0);
        b[i] = a[j];

        }

        // LLamamos al QR
        generateQR(b.join("")  )

        // Almacenamos el token
        return b.join("");
    }

    const generateQR = ( token ) => {

        // Seleccionamos los objetos del boton y canvas
        button.current = document.querySelector('.btn');
        const loginCanvas = document.querySelector('#loginCanvas');

        const url = window.location.href + `login/${ token }/qr`;

        // Generamos el QR
        const setQR = async () => {

        try {
            await QRCode.toCanvas( loginCanvas, url, { scale: 7     } );
        } catch (err) {
            console.error(err);
        }

        }

        // Si el boton no fue generado
        // cambiamos el texto
        if ( !state ) {
            button.current.innerHTML = "RECARGAR QR";
            setState( true );
        }

        // Generamos el QR
        setQR();
    }

    return (
        <>
            <button 
            className='btn btn-lg btn-success mt-3 py-2 px-3'
            onClick={ () => setToken( generateToken() ) }
            >
                GENERAR QR
            </button>
            <canvas id='loginCanvas'></canvas>
        </>
    )

}



