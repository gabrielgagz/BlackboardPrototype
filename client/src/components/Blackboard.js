import React, { useEffect, useState, useRef } from "react";
import { useHistory } from "react-router-dom";
import { useSocket } from "../hooks/useSocket";
import { Modal } from './Modal';
import { v4 as uuidv4 } from 'uuid';
import "../css/blackboard.css";

export const Blackboard = () => {

    // Get token from url
    const urlToken = window.location.pathname
        .replace("/login/", "")
        .replace("/qr", "");

    // Check if user is coming from qr action
    const isQrOn = window.location.href.includes("qr");

    // States
    const [envState, setEnvState] = useState( false );
    const [qrLoad, setQrLoad ] = useState(false);
    const [cleanUp, setCleanUp] = useState( false );

    // Refs
    const lastPoint = useRef();
    const canvasRef = useRef();
    const contextRef = useRef();
    const uuidRef = useRef();

    // Socket custom hook
    const { events, sendEvents, setEvents } = useSocket(urlToken);

    // Modal & actions
    const [modalAction, setModalAction] = useState();
    const modal = Modal( modalAction );

    // Get window size from history
    const history = useHistory();
    const histWidth = history.location.state?.width;
    const histHeight = history.location.state?.height;

    // Set unique id per device
    uuidRef.current = uuidv4();

    useEffect( () => {

        // Canvas environment
        canvasRef.current = document.querySelector("#bCanvas");
        contextRef.current = canvasRef.current.getContext("2d");

        // Window Size (bigger screen resize canvas to match lower screen)
        if ( (window.innerWidth + window.innerHeight) >  ( histWidth + histHeight) ) {

            canvasRef.current.width = histWidth;
            canvasRef.current.height = histHeight;
            canvasRef.current.classList.add("bCanvas");
            
        } else {

            canvasRef.current.width = window.innerWidth;
            canvasRef.current.height = window.innerHeight;

        }

        // Define the path
        contextRef.current.strokeStyle = "white";
        contextRef.current.lineWidth = 2;
        contextRef.current.lineCap = "round";

    },[ envState ]);


    useEffect(() => {

        // Check if user comes from Qr code, if so, send token and window size
        if (isQrOn && !qrLoad) {

            sendEvents([{ token: urlToken, width: window.innerWidth, height: window.innerHeight }]);
            setQrLoad(true);

        }

        // Check if events "cleanup" exists, if so, cleanup all events from server
        if ( events.length > 0 ) {

            // Back to home if user limit is exceeded
            if ( events[events.length -1][0].nousers ) {

                const hiddenButton = document.querySelector('.hiddenButton');
                hiddenButton.click();
                setModalAction('user-alone');

            }
            

            // Back to home if user limit is exceeded
            if ( events[0][0].error ) {

                const hiddenButton = document.querySelector('.hiddenButton');
                hiddenButton.click();
                setModalAction('user-limit');

            }
            
            // Capture the last event, trigered by "CLEAN UP" button
            if (events[events.length -1][0].cleanup) {

                setEvents([]);
                setEnvState(!envState);
                sendEvents([{}]);

            }
            
        }

        // Write
        const draw = ( data ) => {

            // Get coordinate(s) from server
            const { lastPoint, x, y } = data[0];

            if ( lastPoint ) {

                contextRef.current.beginPath();
                contextRef.current.moveTo(lastPoint.x, lastPoint.y);
                contextRef.current.lineTo(x, y);
                contextRef.current.stroke();
                contextRef.current.closePath();

            }
            
        };

        // Actions executed when mouse is moved or touch is detected
        const eventMove = ( event, condition ) => {

            const posX = event.pageX - canvasRef.current.offsetLeft;
            const posY = event.pageY - canvasRef.current.offsetTop;
            

            if ( condition ) {

                if (!lastPoint.current) {

                    lastPoint.current = { x: posX, y: posY };
                    return;

                }
                
                draw([
                    {
                        lastPoint: lastPoint.current,
                        x: posX,
                        y: posY
                    }
                ]); 

                sendEvents([
                    {
                        lastPoint: lastPoint.current,
                        x: posX,
                        y: posY,
                        uuid: uuidRef.current
                    }
                ]);

                lastPoint.current = { x: posX, y: posY };
            }
        };

        // Mouse Wrapper for move events
        const mouseMove = ( event ) => {

            const condition = event.buttons;
            eventMove( event, condition );

        }

        // Touch Wrapper for move events
        const touchMove = ( event ) => {

            const touchEvent = event.changedTouches[0];
            const condition = event.touches;
            eventMove( touchEvent, condition );
            event.preventDefault();

        }

        // Action executed when button or touch is released
        const moveUp = () => {

            lastPoint.current = undefined;

        };

        // Clean up the screen
        if ( cleanUp ) { 
            
            sendEvents([{ cleanup: true }]);
            setCleanUp( false );

        }

        // Get coordinates and draw
        if ( events.length > 0 ) {

            if ( events[0][0].uuid !== uuidRef.current ) {

                events.forEach(draw);
                
            } 
        }

        // Mouse Events
        canvasRef.current.onmousemove = mouseMove;
        canvasRef.current.onmouseup = moveUp;

        // Touch Events
        canvasRef.current.ontouchmove = touchMove;
        canvasRef.current.ontouchend = moveUp;
        canvasRef.current.ontouchcancel = moveUp;
        
    });

    return (
        <div className='blackboard'>
            <canvas id="bCanvas"></canvas>
            <div className='buttons d-grid gap-2 md-block'>
                <button
                    className ='btn btn-primary'
                    type = 'button'
                    onClick = { () => setCleanUp( true ) }
                    >
                    CLEAN UP
                </button>
                <button 
                    className='btn btn-danger'
                    type = 'button'
                    data-bs-toggle="modal"  data-bs-target="#bbModal"
                    onClick = { () => setModalAction( 'user-quit' ) }
                >
                    QUIT
                </button>
                { modal }
                <button 
                type = 'button'
                className = 'hiddenButton'
                data-bs-toggle="modal" data-bs-target="#bbModal"
                >
                </button>
            </div>
        </div>
    );
};