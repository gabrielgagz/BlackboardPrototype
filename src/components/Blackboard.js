import React from 'react';
import '../css/blackboard.css';
import { Canvas } from '../helpers/Canvas';

export const Blackboard = () => {

    Canvas();

    return (
        <div className='blackboard'>
            <Canvas />
            <div className='buttons'>
                <button className='btn btn-primary mb-3 col-sm'>NOTAS</button>
                <button className='btn btn-danger col-sm'>SALIR</button>
            </div>
        </div>
    )
}
