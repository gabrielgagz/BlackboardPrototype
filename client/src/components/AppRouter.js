import React from 'react'
import { Switch, Route, BrowserRouter as Router, Redirect} from 'react-router-dom';
import { Blackboard as BlackboardPage } from './Blackboard';
import { Login as LoginPage } from './Login';

export const AppRouter = () => {
    
    return (
        <Router>
            <>
                <Switch>
                    <Route exact path="/" component={ LoginPage }/>
                    <Route exact path="/login/:token/:qr?" component={ BlackboardPage }/>
                    <Redirect to="/" />
                </Switch>
            </>
        </Router>
    )
}
