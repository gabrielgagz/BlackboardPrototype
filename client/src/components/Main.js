import React from 'react';
import { AppRouter } from './AppRouter';

export const Main = () => {
    return (
        <div className='container-fluid'>
            <AppRouter />
        </div>
    )
}
