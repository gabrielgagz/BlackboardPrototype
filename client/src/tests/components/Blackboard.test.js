import React from 'react';
import { shallow } from 'enzyme';
import { Blackboard } from '../../components/Blackboard';

describe('Blackboard component tests', () => {

    test('Component should be rendered', () => {
        
        const wrapper = shallow( <Blackboard /> );

        expect( wrapper ).toMatchSnapshot();

    });
    
});