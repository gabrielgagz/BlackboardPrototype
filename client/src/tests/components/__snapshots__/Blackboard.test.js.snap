// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blackboard component tests Component should be rendered 1`] = `
<div
  className="blackboard"
>
  <canvas
    id="bCanvas"
  />
  <div
    className="buttons"
  >
    <button
      className="btn btn-primary mb-3 col-sm"
      onClick={[Function]}
    >
      CLEAN UP
    </button>
    <button
      className="btn btn-danger col-sm"
    >
      EXIT
    </button>
  </div>
</div>
`;
